File: FRED/scripts/fred-api.py
# import libraries
import pandas as pd 
import sqlalchemy
import numpy as np
from sqlalchemy import create_engine
import datetime as dt 
import json
import matplotlib.pyplot as plt
import requests
import plotly.express as px
import datetime
from datetime import datetime 
import xmltojson



# class that will interate with FRED API 

# class that will interate with FRED API 

class FredPy:
    
    
    
    def __init__(self, token=None):
        self.token = token
        self.url = "https://api.stlouisfed.org/fred/series/observations" + \
                    "?series_id={seriesID}&api_key={key}&file_type=json" + \
                    "&observation_start={start}&observation_end={end}&units={units}"    
    
    
    
    def set_token(self, token):
        self.token = token
        
    
    
    
    
    def get_series(self, seriesID, start, end, units):
        url_formatted = self.url.format(
            seriesID=seriesID, 
            start=start, 
            end=end,  
            units=units,
            key=self.token # Replace with your API key or get the key from json secret file
            )
        
        response = requests.get(url_formatted)
        
        # if we get successful response, pull the data
        
        if(self.token):
            if (response.status_code == 200):
                data = pd.DataFrame(response.json()['observations'])[['date', 'value']]
                data = data.dropna(subset=['value'])
                data = data[pd.to_numeric(data['value'], errors='coerce').notnull()]\
                    .assign(date = lambda cols: pd.to_datetime(cols['date']))\
                    .assign(value = lambda cols: cols['value'].astype(float))\
                    .rename(columns = {'value' : seriesID})
                    
                
                return data    
            
            
            else:
                raise Exception("Bad response from API, status code = {}".format(response.status_code))
        else:
            raise Exception("No API key provided")    

File: FRED/scripts/fred-api.R

# install and load necessary packages/libraries

packages <- c("httr2", "jsonlite", "tidyverse", "glue")

# Loop through each package
for (pkg in packages) {
  # Check if the package is already installed
  if (!require(pkg, character.only = TRUE)) {
    # Install the package using renv::install if not already installed
    renv::install(pkg)
    # Load the package after installation
    library(pkg, character.only = TRUE)
  }
}


library(httr2)
library(jsonlite)
library(glue)
library(tidyverse)
library(purrr)  # For `pluck`
library(dplyr)  # For `glimpse`



# write a function that will use httr2 package


get_fred_series <- function(seriesID, start, end, units, api_key) {
  
  base_url <- "https://api.stlouisfed.org/fred/series/observations"
  
  
  # Construct the request
  response <- httr2::request(base_url) |>
    httr2::req_url_query(
      series_id = seriesID,
      api_key = api_key,
      file_type = "json",
      observation_start = start,
      observation_end = end,
      units = units
    ) |>
    httr2::req_perform()
  
  # Check if request was successful
  if (httr2::resp_status(response) == 200) {
    
    response_json <- 
      response |>
      httr2::resp_body_json() |>
      pluck('observations') |> 
      map_dfr(
        \(x) {
          tibble(
            date = x |> pluck('date'),
            value = x |> pluck('value')
          )
        }
      ) |> 
      setNames(c("date", seriesID))
    
    return(response_json)
    
  } else {
    
    # Message if API call is unsuccessful
    message("Request failed with status: ", httr2::resp_status(response))
    return(NULL)
  }
}



get_more_series <- 
  function(seriesIDs, start, end, units, api_key) {
    
    
    # initialise an empty list that will store the series 
    
    series_list <- lapply(seriesIDs, function(seriesID) {
      
      # get series data 
      
      series_data <- get_fred_series(seriesID, start, end, units, api_key)
      
      # now let us rename our columns accordingly; to match seriesID for ease of interpretation
      
      names(series_data)[2] <- seriesID
      return(series_data)
    })
    
    
    # join these datasets into one data frame
    
    
    combined_data <- purrr::reduce(
      series_list,
      dplyr::full_join, by = 'date')
    
    combined_data <- 
      combined_data |> 
      dplyr::mutate(date = lubridate::as_date(date))
    
    combined_data[-1] <- lapply(combined_data[-1], as.numeric)
    
    return(combined_data)
  }

